Fixes #38673 - dev docs - ui tests

MariaAga opened the pull request.

ShimShtein commented on developer_docs/ui-testing-guidelines.asciidoc:
```diff
@@ -0,0 +1,74 @@
+Some mocked files are already in `\\__mocks__` folder, but dont define all the functions from the original file, so they will show up as undefined.
+- Tests files should be named like `*.test.js`, and should be in the same folder as the component, or in the `\\__tests__` folder.
+- Tests can mock the API in a few ways:
+  * By adding the mock data to the store using `redux-mock-store`
```
`redux-mox-store` is already [deprecated](https://github.com/reduxjs/redux-mock-store?tab=readme-ov-file#deprecation-notice). Are we sure we want to document an already deprecated technology? They point to [redux documentation](https://redux.js.org/usage/writing-tests) for more info on how to test redux enabled components.

MariaAga replied:
Thanks for the note, I'll add one test that uses redux correctly to this pr
Should we encourage people in the guidelines to use AI ?

ShimShtein replied:
I think it's up to the developer. I would ask that if a piece of code was generated by AI, the developer should notify about it in the PR.

ShimShtein commented on developer_docs/ui-testing-guidelines.asciidoc:
```diff
@@ -0,0 +1,74 @@
+
+For React Testing Library tests, you can use:
+
+* jest.useFakeTimers(); and then jest.advanceTimersByTime(1000);
```
BTW: to avoid `act` errors, you have to wrap the `advanceTimersByTime` calls in `act`
``` js
      await act(async () => jest.advanceTimersByTime(1000));
```

MariaAga commented:
Updated the docs to show how to setup redux, created an example and a generic function.
Had to change the `/API` mocks to mock only `/API/API` so the test setup wont be broken. mocked axios globally as accidently using it will result in a network error.
The `OperatingSystem.test.js` was written by AI and edited and reviewed by me

ShimShtein commented on developer_docs/ui-testing-guidelines.asciidoc:
```diff
@@ -0,0 +1,85 @@
+You can read more about React Testing Library in https://testing-library.com/docs/guiding-principles/[React Testing Guiding Principles].
+Test examples:
+
+- webpack/assets/javascripts/react_app/routes/common/PageLayout/PageLayout.test.js
```
I think it's worth making these links: 
```suggestion
- [webpack/assets/javascripts/react_app/routes/common/PageLayout/PageLayout.test.js](../webpack/assets/javascripts/react_app/routes/common/PageLayout/PageLayout.test.js)
```

ShimShtein commented on developer_docs/ui-testing-guidelines.asciidoc:
```diff
@@ -0,0 +1,85 @@
+- webpack/assets/javascripts/react_app/components/Loading/\\__tests__/Loading.test.js
+
+- Any test that ran by jest (`npm run test`) might use mocked functions.
+You can set a mock per test, or globally by adding the mock to `\\__mocks__` folder.
```
Do you mean a single folder, as in https://github.com/theforeman/foreman/tree/1b13ec8dd3c67368eba1597711ed108d49c2140e/webpack/assets/javascripts/react_app/common/__mocks__ or it means we can create a `__mocks__` folder under any component and the folder will be considered the scope for the mocks?

ShimShtein commented on developer_docs/ui-testing-guidelines.asciidoc:
````diff
@@ -0,0 +1,85 @@
+
+```
+  * By mocking the selectors return values
+  * By mocking the API calls directly
````
Do we have a recommendation which method is preferred, or what are the conditions to prefer one to the other?

MariaAga replied:
Not really, I think people find different things more easy to write/how the code is built

ShimShtein replied:
So let's create a preferred option: I would say it's API mock + predefined state fixture. I think we shouldn't mock the selectors, since we don't test them anywhere else.

MariaAga replied:
Changed the guidelines

ShimShtein commented on developer_docs/ui-testing-guidelines.asciidoc:
```diff
@@ -0,0 +1,85 @@
+For Capybara tests, you can use:
+
+* wait_for_ajax
+* or other wait_for_... methods
```
In the vast majority of the tests we are using `wait_for_ajax`. I don't think we should mention other methods.

ShimShtein commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/__tests__/components/OperatingSystem.test.js:
```diff
@@ -1,12 +1,170 @@
+  };
 
-jest.mock('react-redux');
+  return render(
```
Now you have `rtlHelpers.renderWithStore` helper: https://github.com/theforeman/foreman/blob/1b13ec8dd3c67368eba1597711ed108d49c2140e/webpack/assets/javascripts/react_app/common/rtlTestHelpers.js#L18

MariaAga replied:
Update rtlHelpers to use a real store and used it

ShimShtein reviewed (commented):
Can you please fix the commit message? You have refactored the `OperatingSystem` component test, but it is not obvious from the commit message.
I may even argue that these two should be separated into two PRs.

MariaAga commented:
>I may even argue that these two should be separated into two PRs.

The OS test is because it mocks redux, so it breaks with this PR so I rewrote it to not mock redux

MariaAga commented:
Updated the commit message to be long

MariaAga commented:
the real store (compared to the mocked store) has not `getActions` so I mocked the actions and checked if they are called instead

ShimShtein commented on developer_docs/ui-testing-guidelines.asciidoc:
````diff
@@ -0,0 +1,90 @@
+  expect(screen.getByText('text')).toBeInTheDocument();
+};
+```  
+  * By mocking the API calls directly
````
For me GH preview renders the bullet points on an incorrect indentation level. Notice the `By mocking the API calls directly` indentation:


<img width="646" height="591" alt="image" src="https://github.com/user-attachments/assets/9865bf0a-7497-4173-a5e7-68442607927d" />

ShimShtein commented on developer_docs/ui-testing-guidelines.asciidoc:
```diff
@@ -0,0 +1,90 @@
+For plugins, the mock folder for Foreman core is in `webpack/__mocks__/foremanReact/`
+- Tests files should be named like `*.test.js`, and should be in the same folder as the component, or in the `\\__tests__` folder.
+- Tests can mock the API in a few ways:
+  * By adding the mock data to the store using `rtlHelpers.renderWithStore`. `renderWithStore` will create a Redux store with the mock data using an initial state mock. You can either add to the mock in `/testInitialReduxStore.js` or send a custom initial state as a second argument.
```
```suggestion
  * By adding the mock data to the store using `rtlHelpers.renderWithStore`. `renderWithStore` will create a Redux store with the mock data using an initial state mock. You can either add to the global mock in `/testInitialReduxStore.js` or send a custom initial state as a second argument.
```

ShimShtein commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/__tests__/components/OperatingSystem.test.js:
```diff
@@ -1,12 +1,138 @@
-import { testComponentSnapshotsWithFixtures } from '../../../../../common/testHelpers';
+import React from 'react';
+import { screen, fireEvent, waitFor,act } from '@testing-library/react';
```
```suggestion
import { screen, fireEvent, waitFor, act } from '@testing-library/react';
```

ShimShtein commented on developer_docs/ui-testing-guidelines.asciidoc:
```diff
@@ -0,0 +1,92 @@
+Tests files should be named like `*.test.js`, and should be in the same folder as the component, or in the `\\__tests__` folder.
+Tests can mock the API in a few ways:
+
+* By adding the mock data to the store using `rtlHelpers.renderWithStore`. `renderWithStore` will create a Redux store with the global mock data using an initial state mock. You can either add to the mock in `/testInitialReduxStore.js` or send a custom initial state as a second argument.
```
I suppose the main usage will be something like this:
``` js
mySpecialStore = {
  ...initDefaultStore(),
  myProp1: 1,
  myProp2: 2,
};

renderWithStore(
  <MyComponent />,
  mySpecialStore
);
```
Maybe we should change the `renderWithStore` function to accept only the changes for the default store, and not the whole store to make the code look like this:

``` js
renderWithStore(
  <MyComponent />,
  {
    myProp1: 1,
    myProp2: 2,
  }
);
```
This should improve the readability.

MariaAga replied:
done

MariaAga committed:
Fixes #38673 - dev docs - ui tests and js test updates

Updates JS OS test to not mock redux, and other tests to mock only the API function they are importing. modifys test helpers to use real redux store instead of a mock. added empty mocks for axios and api requests for all JS tests.
